### PR TITLE
[cuda] modify the fused adam, support hybrid of fp16 and fp32

### PR DESCRIPTION
**Pull request by Gy-Lu**
now our fused adam supports the hybrid of p and d dtype.

### CODE REVIEW

File: tests/test_optimizer/unittest_fused_adam.py [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+        y = model_copy(data_copy[i])
+        loss = ((label_copy[i] - y) ** 2).sum()
+        torch_optim.zero_grad()
+        loss.backward()
+        torch_optim.step()
+
+    assert len(optim.param_groups[0]['params']) == len(torch_optim.param_groups[0]['params'])
+    max_diff = 0
+    for i in range(len(optim.param_groups[0]['params'])):
+        diff = torch.max(torch.abs(optim.param_groups[0]['params'][i]-(torch_optim.param_groups[0]['params'][i])))
```
Comment (ver217) [inline tests/test_optimizer/unittest_fused_adam.py:60]: Is it possible to use `torch.allclose()`?

Comment (Gy-Lu) [reply tests/test_optimizer/unittest_fused_adam.py:60]: yes, thanks for your advice.

File: tests/test_optimizer/unittest_fused_adam.py [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+        y = model(data[i])
+        loss = ((label[i] - y) ** 2).sum()
+        optim.zero_grad()
+        loss.backward()
+        if p_dtype != g_dtype:
+            for i in range(len(optim.param_groups[0]['params'])):
+                optim.param_groups[0]['params'][i].grad.data = optim.param_groups[0]['params'][i].grad.data.to(g_dtype)
+        optim.step()
+
+    for i in range(len(data_copy)):
```
Comment (ver217) [inline tests/test_optimizer/unittest_fused_adam.py:50]: Can we extract this part and write a function which both optimizers can use?

Comment (Gy-Lu) [reply tests/test_optimizer/unittest_fused_adam.py:50]: Yes, but I am not sure if it is necessary. If we write a function to reuse, there would be some code to judge the states and a lot of parameters to pass.

File: tests/test_optimizer/unittest_cpu_adam.py [hunk 3]
```diff
@@ -121,7 +88,7 @@ def check_res(
             p_grad.view(-1),    # fp32 grad
             exp_avg.view(-1),
             exp_avg_sq.view(-1),
-            loss_scale,
+            -1,
```
Comment (feifeibear) [inline tests/test_optimizer/unittest_cpu_adam.py:91]: why we set loss_scale as -1?

File: tests/test_optimizer/unittest_fused_adam.py [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+        loss.backward()
+        torch_optim.step()
+
+    assert len(optim.param_groups[0]['params']) == len(torch_optim.param_groups[0]['params'])
+    # we set the torch adam's dtype always being fp32
+    # when p and g both are fp16, there is a great loss
+    # so we make a loose threshold
+    
+    #threshold = 1e-3
+    print(adamw, p_dtype, g_dtype)
```
Comment (feifeibear) [inline tests/test_optimizer/unittest_fused_adam.py:61]: remove this line

File: tests/test_optimizer/unittest_fused_adam_kernel.py [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+                m_copy,
+                v_copy,
+                -1,
+                adamw,
+            )
+        
+        if torch.isnan(p).any() or torch.isnan(p_copy).any():
+            count += 1
+            continue
+        assert count < 200, "too many nans"
```
Comment (feifeibear) [inline tests/test_optimizer/unittest_fused_adam_kernel.py:97]: did we allow nan in this check?

Comment (Gy-Lu) [reply tests/test_optimizer/unittest_fused_adam_kernel.py:97]: yes, for fp16, nan seems unavoidable, especially when there are convertions.

File: tests/test_optimizer/unittest_cpu_adam.py (hunk 3 above)
Comment (Gy-Lu) [reply tests/test_optimizer/unittest_cpu_adam.py:91]: oh, I just forget to change it...
it should be -1 or 2**5

Comment (Gy-Lu) [reply tests/test_optimizer/unittest_cpu_adam.py]: Have fixed it.

File: tests/test_optimizer/unittest_fused_adam.py (hunk 4 above)
Comment (Gy-Lu) [reply tests/test_optimizer/unittest_fused_adam.py]: Done

**Review by feifeibear — APPROVED** [review]

**Pull request merged**